src/render.rs

```diff
@@ -0,0 +1,140 @@
+use matrix_sdk::events::room::{
+    encrypted::EncryptedEvent,
+    member::{MemberEvent, MembershipState},
+    message::{
+        AudioMessageEventContent, EmoteMessageEventContent,
+        FileMessageEventContent, ImageMessageEventContent, MessageEvent,
+        MessageEventContent, NoticeMessageEventContent,
+        TextMessageEventContent, VideoMessageEventContent,
+    },
+};
+
+/// This trait describes events that can be rendered in the weechat UI
+pub(crate) trait RenderableEvent {
+    /// Convert the event into a string that will be displayed in the UI.
+    /// The displayname is taken as a parameter since it cannot be calculated from the event
+    /// context alone.
+    fn render(&self, displayname: &str) -> String;
+}
+
+impl RenderableEvent for EncryptedEvent {
+    // TODO: this is not implemented yet
+    fn render(&self, displayname: &str) -> String {
+        format!("{}\t{}", displayname, "Unable to decrypt message")
+    }
+}
+
+impl RenderableEvent for MemberEvent {
+    fn render(&self, displayname: &str) -> String {
+        let operation = match self.content.membership {
+            MembershipState::Join => "joined",
+            MembershipState::Leave => "left",
+            MembershipState::Ban => "banned",
+            MembershipState::Invite => "invited",
+            MembershipState::Knock => "knocked on", // TODO
+        };
+        format!(
+            "{} ({}) has {} the room",
+            displayname, self.state_key, operation
+        )
+    }
+}
+
+impl RenderableEvent for MessageEvent {
+    fn render(&self, displayname: &str) -> String {
```

Comment:
I wonder if we should return the message and the tags here. (tags are metadata that are attached to the event, you can see them with `/debug tags`). For example we'll always want the user_id of the sender, the event_id,  perhaps some other stuff in the tags.

Reply:
I went with adding `fn sender(&self) -> &UserId;` to the `RenderableEvent` trait. If needed, we can also add `fn event_id(&self) -> &EventId;`.